File: src/main/java/tech/sirwellington/alchemy/generator/BinaryGenerators.java
/*
 * Copyright 2025 Sir Wellington.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

 
package tech.sirwellington.alchemy.generator;


import java.nio.ByteBuffer;
import org.apache.commons.lang3.RandomUtils;
import tech.sirwellington.alchemy.annotations.access.NonInstantiable;
import tech.sirwellington.alchemy.annotations.designs.patterns.StrategyPattern;

import static tech.sirwellington.alchemy.annotations.designs.patterns.StrategyPattern.Role.CONCRETE_BEHAVIOR;
import static tech.sirwellington.alchemy.generator.ChecksJ.checkThat;

/**
 * {@link AlchemyGenerator Alchemy Generators} for raw binary (`byte[]`).
 * 
 * @author SirWellington
 */
@NonInstantiable
@StrategyPattern(role = CONCRETE_BEHAVIOR)
public class BinaryGenerators {

    private BinaryGenerators() throws IllegalAccessException  {
        throw new IllegalAccessException("Cannot instantiate this class");
    }
    
    /**
     * Generates binary of the specified length.
     * @param length The size of the byte arrays created.
     * @return A binary generator.
     * @throws IllegalArgumentException If `length < 0`.
     * @see #byteBuffers(int) 
     */
    static AlchemyGenerator<byte[]> binary(int length) {
        checkThat(length >= 0, "length must be >= 0");
        return () -> RandomUtils.secure().randomBytes(length);
    }
    
    /**
     * Generates a {@link ByteBuffer} of the specified size.
     * @param size The size of the byte buffers produced.
     * @return A {@link ByteBuffer} of the specified size.
     * @throws IllegalArgumentException If `length < 0`.
     * @see #binary(int) 
     */
    static AlchemyGenerator<ByteBuffer> byteBuffers(int size) throws IllegalArgumentException {
        checkThat(size > 0, "length must be >= 0");
        AlchemyGenerator<byte[]> delegate = binary(size);
        return () -> ByteBuffer.wrap(delegate.get());
    }
    
    /**
     * Generates a single `byte`.
     * @return A {@link AlchemyGenerator} that produces a single byte.
     */
    static AlchemyGenerator<Byte> bytes() {
        return () -> RandomUtils.secure().randomBytes(1)[0];
    }
}

File: src/main/java/tech/sirwellington/alchemy/generator/BinaryGenerators.kt
/*
 * Copyright 2017 SirWellington Tech.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package tech.sirwellington.alchemy.generator

import org.apache.commons.lang3.RandomUtils
import org.slf4j.LoggerFactory
import tech.sirwellington.alchemy.annotations.access.NonInstantiable
import tech.sirwellington.alchemy.annotations.designs.patterns.StrategyPattern
import tech.sirwellington.alchemy.annotations.designs.patterns.StrategyPattern.Role.CONCRETE_BEHAVIOR
import java.lang.IllegalArgumentException
import java.nio.ByteBuffer

/**
 * [Alchemy Generators][AlchemyGenerator] for raw binary (`byte[]`).
 *
 * @author SirWellington
 */
@NonInstantiable
@StrategyPattern(role = CONCRETE_BEHAVIOR)
class BinaryGenerators
@Throws(IllegalAccessException::class)
internal constructor()
{

    init
    {
        throw IllegalAccessException("cannot instantiate this class")
    }

    companion object
    {

        private val LOG = LoggerFactory.getLogger(BinaryGenerators::class.java)

        /**
         * Generates binary of the specified length
         *
         * @param length The size of the byte arrays created.
         *
         *
         * @return A binary generator
         *
         * @throws IllegalArgumentException If `length < 0`.
         */
        @JvmStatic
        @Throws(IllegalArgumentException::class)
        fun binary(length: Int): AlchemyGenerator<ByteArray>
        {
            checkThat(length >= 0, "length must be >= 0")

            return AlchemyGenerator { RandomUtils.nextBytes(length) }
        }

        /**
         * Generates a [ByteBuffer] of the specified length.
         *
         * @param size The desired size of the Byte Buffer.
         *
         * @return
         *
         * @throws IllegalArgumentException If `size < 0`.
         */
        @Throws(IllegalArgumentException::class)
        @JvmStatic
        fun byteBuffers(size: Int): AlchemyGenerator<ByteBuffer>
        {
            checkThat(size >= 0, "size must be at least 0")

            val delegate = binary(size)

            return AlchemyGenerator {
                val binary = delegate.get()
                ByteBuffer.wrap(binary)
            }
        }

        @JvmStatic
        fun bytes(): AlchemyGenerator<Byte>
        {
            val delegate = binary(1)

            return AlchemyGenerator {
                val array = delegate.get()
                checkThat(array != null && array.isNotEmpty(), "no Byte available to return.")

                array!![0]
            }
        }
    }

}
